File: Java/Hasta Entraga 3/irrgarten/src/irrgarten/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 *
 * @author alejandro
 */
public class Game {
    // Atributos de clase
    private static final int MAX_ROUNDS = 10;
    private static final int MAX_PLAYERS = 4; // Número de jugadores máximos
    private static final int MAX_MONSTERS = 10; // Número de monstruos máximos
    private static final int ROWS = 10; // Filas por defecto - Preguntar
    private static final int COLS = 10; // Comlumnas por defecto - Preguntar
    
    // Atributos de instancia
    private int currentPlayerIndex;
    private String log;
    
    private Labyrinth labyrinth;
    private ArrayList<Player> players;
    private ArrayList<Monster> monsters;

    // Constructor
    public Game(int currentPlayerIndex) {
        this.currentPlayerIndex = currentPlayerIndex;
        this.log = " ";
        
        // Inicializo los players
        this.players = new ArrayList<>();
        
        for(int i=0; i < MAX_PLAYERS; i++){
            Player newPlayer = new Player(i, Dice.randomIntelligence(), Dice.randomStrength());
            this.players.add(newPlayer);
        }
        
        // Inicializo los monstruos
        this.monsters  = new ArrayList<>();
        
        // Asumo los valores del laberinto ------ Preguntar al profesor
        int rows = ROWS;
        int cols = COLS;
        int exitR = Dice.randomPos(rows);
        int exitC = Dice.randomPos(cols);
        
        // Inicializo el laberinto
        this.labyrinth = new Labyrinth(rows,cols,exitR,exitC);
        
        // Lamo al método que te configura el laberinto
        this.configureLabyrinth();
        
        this.labyrinth.spreadPlayers(players);
    }
    
    
    // Devuelve true si hay un ganador
    public boolean finished(){
        return labyrinth.haveAWinner();
    }
    
    // Método completado
    public boolean nextStep(Directions preferredDirection){
        this.log = "";
        boolean dead = players.get(currentPlayerIndex).dead();
        
        if(!dead){
            Directions direction = actualDirection(preferredDirection);
            
            if(direction != preferredDirection){
                logPlayerNoOrders();
            }
            
            Monster monster = labyrinth.putPlayer(direction, players.get(currentPlayerIndex));
        
            if(monster == null){
                logNoMonster();
            } else{
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
        } else{
            manageResurrection();
        }
        
        boolean endGame = finished();
        
        if(!endGame){
            nextPlayer();
        }
        
        return endGame;
    }
    
    // Devuelve toda la informacición actual del estado del juegoó
    public GameState getGameState(){
        String laberinto = this.labyrinth.toString();
        
        String playersString = "";
        for(int i= 0; i < MAX_PLAYERS;i++){
            playersString += this.players.get(i).toString() + "\n";
        }
        
        String monsterString = " ";
        for(int i=0; i < MAX_MONSTERS; i++){
            monsterString += this.monsters.get(i).toString() + "\n";
        }
        
        return new GameState(laberinto,playersString,monsterString,this.currentPlayerIndex,this.finished(), this.log);
    }
    
    // Configura el laberinto añadiendo bloques de obstáculos y monstruos.
    private void configureLabyrinth(){
    // Añado bloques para rodear el escenario
        this.labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, COLS); // Pared superior
        this.labyrinth.addBlock(Orientation.HORIZONTAL, ROWS - 1, 0, COLS); // Pared inferior
        this.labyrinth.addBlock(Orientation.VERTICAL, 0, 0, ROWS); // Pared izquierda
        this.labyrinth.addBlock(Orientation.VERTICAL, 0, COLS - 1, ROWS); // Pared derecha
        

        this.labyrinth.addBlock(Orientation.HORIZONTAL, 2, 2, 4); 
        this.labyrinth.addBlock(Orientation.VERTICAL, 4, 5, 4);   
        this.labyrinth.addBlock(Orientation.VERTICAL, 6, 2, 3);
        this.labyrinth.addBlock(Orientation.HORIZONTAL, 3, 7, 2);
        
        // Añado monstruos
        for(int i = 0; i < MAX_MONSTERS; i++){
            Monster monster = new Monster("Monster #" + i, Dice.randomIntelligence(), Dice.randomStrength());
            
            int r = Dice.randomPos(ROWS);
            int c = Dice.randomPos(COLS);
            
            this.labyrinth.addMonster(r, c, monster);
            
            this.monsters.add(monster);
        }
    }
    
    // Actualiza los dos atributos que indican el jugador (current*) con el turno pasando al siguiente jugador.
    private void nextPlayer(){
        // Uso el módulo para que sea ciclicoó
        this.currentPlayerIndex = (this.currentPlayerIndex + 1) % this.players.size();
    }
    
    // Método completado
    private Directions actualDirection(Directions preferredDirection){
        int currentRow = players.get(currentPlayerIndex).getRow();
        int currentCol = players.get(currentPlayerIndex).getCol();
        
        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);
        
        Directions output = players.get(currentPlayerIndex).move(preferredDirection, validMoves);
        
        return output;
    }
    
    // Método completado
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner;
        winner = GameCharacter.PLAYER;
        
        float playerAttack = players.get(currentPlayerIndex).attack();
        boolean lose = monster.defend(playerAttack);
        
        while(!lose && (rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;
            
            float monsterAttack = monster.attack();
            
            lose = players.get(currentPlayerIndex).defend(monsterAttack);
            
            if(!lose){
                playerAttack = players.get(currentPlayerIndex).attack();
                
                winner = GameCharacter.PLAYER;
                
                lose = monster.defend(playerAttack);
            }
        }
        
        logRounds(rounds,MAX_ROUNDS);
        
        return winner;         
    }
    
    // Método completado
    private void manageReward(GameCharacter winner){
        if(winner == GameCharacter.PLAYER){
            players.get(currentPlayerIndex).receiveReward();
            logPlayerWon();
        } else{
            logMonsterWon();
        }
    }
    
    // Método completado
    private void manageResurrection(){
        boolean resurrect = Dice.resurrectPlayer();
        
        if(resurrect){
            players.get(currentPlayerIndex).resurrect();
            logResurrected();
        } else{
            logPlayerSkipTurn();
        }
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el jugador ha ganado el combate.
    private void logPlayerWon(){
        this.log += " ¡Victoria! \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el monstruo ha ganado el combate.
    private void logMonsterWon(){
        this.log += " ¡Monster Won! \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el jugador ha resucitado
    private void logResurrected(){
        this.log += " ¡El jugador ha resucitado! \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el jugador ha perdido el turno por estar muerto.
    private void logPlayerSkipTurn(){
        this.log += " ¡El jugador perdió el turno! \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el jugador no ha seguido las instrucciones del jugador humano (no fue posible).
    private void logPlayerNoOrders(){
        this.log += " ¡No se puede realizar esa acción! \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el mensaje
    // indicando que el jugador se ha movido a una celda vacía o no le ha sido posible moverse.
    private void logNoMonster(){
        this.log += " ¡Se ha movido a una casilla vacia, no se puede realizar dicha acción \n";
    }
    
    // Añade al final del atributo log (concatena cadena al final) el
    // mensaje que se han producido rounds de max rondas de combate. También añade el indicador de
    // nueva línea al final.
    private void logRounds(int rounds, int max){
        this.log += " ¡Se han producido " + rounds + " de " + max + " rondas! \n";
    }
    
    
    
}

File: Ruby/Hasta Entrega 3/Irrgarten/Irrgarten/game.rb
module Irrgarten
  class Game
    # Atributos de clase
    @@max_rounds = 10
    @@max_players = 4
    @@max_monsters = 10
    @@rows = 10
    @@cols = 10

    # Atributos de instancia
    def initialize(current_player_index)
      @current_player_index = current_player_index
      @log = " "

      # Inicializo los players
      @players = []
      (0...@@max_players).each do |i|
        new_player = Player.new(i, Dice.random_intelligence, Dice.random_strength)
        @players << new_player
      end

      # Inicializo los monstruos
      @monsters = []

      @rows = @@rows
      @cols = @@cols
      @exitR = Dice.random_pos(@rows)
      @exitC = Dice.random_pos(@cols)

      # Inicializo el laberinto
      @labyrinth = Labyrinth.new(@rows, @cols, @exitR, @exitC)

      # Llamo al método que configura el laberinto
      configure_labyrinth

      @labyrinth.spread_players(@players)
    end

    # Devuelve true si hay un ganador
    def finished
      @labyrinth.have_a_winner
    end

    # Metodo completado
    def next_step(preferred_direction)
      @log = ""
      current_player = @players[@current_player_index]
      dead = current_player.dead?

      if !dead
        direction = actual_direction(preferred_direction)

        if direction != preferred_direction
          log_player_no_orders
        end

        monster = @labyrinth.put_player(direction, current_player)

        if monster.nil? # .nil? es el '== null' de Ruby
          log_no_monster
        else
          winner = combat(monster)
          manage_reward(winner)
        end
      else
        manage_resurrection
      end

      end_game = finished

      if !end_game
        next_player
      end

      end_game
    end

    # Devuelve toda la información actual del estado del juego
    def get_game_state
      laberinto = @labyrinth.to_s

      players_string = ""
      @players.each do |player|
        players_string += "#{player.to_s}\n"
      end

      monster_string = " "
      @monsters.each do |monster|
        monster_string += "#{monster.to_s}\n"
      end

      GameState.new(laberinto, players_string, monster_string, @current_player_index, finished, @log)
    end

    # Configura el laberinto añadiendo bloques de obstáculos y monstruos.
    private def configure_labyrinth
      # Añado bloques
      @labyrinth.add_block(Orientation::HORIZONTAL, 0, 0, @@cols)        # Pared superior
      @labyrinth.add_block(Orientation::HORIZONTAL, @@rows - 1, 0, @@cols)  # Pared inferior
      @labyrinth.add_block(Orientation::VERTICAL, 0, 0, @@rows)        # Pared izquierda
      @labyrinth.add_block(Orientation::VERTICAL, 0, @@cols - 1, @@rows)  # Pared derecha

      # Añado monstruos
      @@max_monsters.times do |i|
        monster = Monster.new("Monster ##{i}", Dice.random_intelligence, Dice.random_strength)

        r = Dice.random_pos(@@rows)
        c = Dice.random_pos(@@cols)

        @labyrinth.add_monster(r, c, monster)

        @monsters << monster
      end
    end

    # Actualiza los dos atributos que indican el jugador (current*) con el turno pasando al siguiente jugador.
    private def next_player
      # Uso el módulo para que sea cíclico
      @current_player_index = (@current_player_index + 1) % @players.size
    end

    # Metodo completado
    private def actual_direction(preferred_direction)
      current_player = @players[@current_player_index]

      # (Asumimos que Player tiene attr_reader :row y :col)
      current_row = current_player.row
      current_col = current_player.col

      valid_moves = @labyrinth.valid_moves(current_row, current_col)

      output = current_player.move(preferred_direction, valid_moves)

      output
    end

    # Metodo completado
    private def combat(monster)
      rounds = 0
      winner = GameCharacter::PLAYER
      current_player = @players[@current_player_index]

      player_attack = current_player.attack
      lose = monster.defend(player_attack)

      # Usamos la variable de clase @@max_rounds
      while !lose && (rounds < @@max_rounds)
        winner = GameCharacter::MONSTER
        rounds += 1

        monster_attack = monster.attack
        lose = current_player.defend(monster_attack)

        if !lose
          player_attack = current_player.attack
          winner = GameCharacter::PLAYER
          lose = monster.defend(player_attack)
        end
      end

      log_rounds(rounds, @@max_rounds)

      winner
    end

    # Metodo completado
    private def manage_reward(winner)
      if winner == GameCharacter::PLAYER
        @players[@current_player_index].receive_reward
        log_player_won
      else
        log_monster_won
      end
    end

    # Metodo completado
    private def manage_resurrection
      resurrect = Dice.resurrect_player

      if resurrect
        @players[@current_player_index].resurrect
        log_resurrected
      else
        log_player_skip_turn
      end
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el jugador ha ganado el combate.
    private def log_player_won
      @log += " ¡Victoria! \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el monstruo ha ganado el combate.
    private def log_monster_won
      @log += " ¡Monster Won! \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el jugador ha resucitado
    private def log_resurrected
      @log += " ¡El jugador ha resucitado! \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el jugador ha perdido el turno por estar muerto.
    private def log_player_skip_turn
      @log += " ¡El jugador perdió el turno! \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el jugador no ha seguido las instrucciones del jugador humano (no fue posible).
    private def log_player_no_orders
      @log += " ¡No se puede realizar esa acción! \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el mensaje
    # indicando que el jugador se ha movido a una celda vacía o no le ha sido posible moverse.
    private def log_no_monster
      @log += " ¡Se ha movido a una casilla vacia, no se puede realizar dicha acción \n"
    end

    # Añade al final del atributo log (concatena cadena al final) el
    # mensaje que se han producido rounds de max rondas de combate. También añade el indicador de
    # nueva línea al final.
    private def log_rounds(rounds, max)
      # En Ruby es más idiomático usar interpolación de strings
      @log += " ¡Se han producido #{rounds} de #{max} rondas! \n"
    end

  end
end